mirascope/base/extractors.py

```diff
@@ -303,42 +373,71 @@ async def _stream_async(
+            tool_stream = tool_stream_type.from_async_stream(stream, allow_partial=True)
+            try:
+                yielded = False
+                async for partial_tool in tool_stream:
+                    extracted_schema = self._extract_schema(
+                        partial_tool, self.extract_schema, return_tool, response=None
+                    )
+                    if extracted_schema is None:
+                        break
+                    yielded = True
+                    yield extracted_schema
+
+                if not yielded:
+                    raise AttributeError("No tool found in the completion.")
+            except (AttributeError, ValueError, ValidationError):
+                raise
+
+        if isinstance(retries, int):
+            retries = AsyncRetrying(stop=stop_after_attempt(retries))
         try:
-            yielded = False
-            async for partial_tool in tool_stream:
-                extracted_schema = self._extract_schema(
-                    partial_tool, self.extract_schema, return_tool, response=None
-                )
-                if extracted_schema is None:
-                    break
-                yielded = True
-                yield extracted_schema
-
-            if not yielded:
-                raise AttributeError("No tool found in the completion.")
-        except (AttributeError, ValueError, ValidationError) as e:
-            if retries > 0:
-                logging.info(f"Retrying due to exception: {e}")
-                # TODO: include failure in retry prompt.
-                async for partial_tool in self._stream_async(
-                    call_type, tool_type, tool_stream_type, retries - 1, **kwargs
-                ):
-                    yield partial_tool
-            raise  # re-raise if we have no retries left
-
-    def _generate_temp_call(self, call_type: Type[BaseCallT]) -> Type[BaseCallT]:
+            error_message = None
+            async for attempt in retries:
+                with attempt:
+                    try:
+                        async for partial_tool in _stream_attempt_async(
+                            call_type,
+                            tool_type,
+                            tool_stream_type,
+                            error_message,
+                            **kwargs,
+                        ):
+                            yield partial_tool
+                    except (AttributeError, ValueError, ValidationError) as e:
+                        error_message = e
+                        logger.error(f"Retrying due to exception: {e}")
+                        raise
+        except RetryError as e:
+            raise e
+
+    def _generate_temp_call(
+        self, call_type: Type[BaseCallT], error_message: Optional[str] = None
+    ) -> Type[BaseCallT]:
         """Returns a `TempCall` generated using the extractors definition."""
+        _prompt_template = self.prompt_template
+        if error_message:
```

Comment:
I think we should consider using `error_messages` as an array (or dictionary keys for faster check/access) and continuously adding for each retry (and we can reduce the number of errors included by only including those of different types).

Thoughts?

Reply:
okay so a set?

Reply:
well, sets aren't ordered, and we likely want the errors in order so the LLM can identify the order of failures, so best to actually just use a dictionary and then grab the set from the keys, right?

Reply:
e.g. checking `if thing in dict` will have O(1) access like a set, but then when we want to inject them we can do `list(dict.keys())` and ensure they are in order